### ISSUE TITLE
Python in the browser as a solution to pickling

### ISSUE THREAD

**Issue by lidavidm** [sympy/sympy-live#83]
Pickling is a problem for SymPy Live as it can lead to strange errors (not everything is picklable); running Python in the browser could avoid this. There are three options here:
- CPython via Native Client
  - Pros: full CPython
  - Cons: Chrome only, large size
- CPython via Emscripten
  - Pros: full CPython
  - Cons: slower than CPython, large size, seems to hang/crash browsers (when I tried)
- PyPy via [PyPy.js](https://rfk.id.au/blog/entry/pypy-js-faster-than-cpython/)
  - Pros: relatively small (only 32MB JavaScript)
  - Cons: relatively new, buggy (`import random` leads to an error)

I'm leaving this here mostly to record ideas/options.

**Comment by stefnotch** [comment]
It is possible to run [Sympy in the browser using Pyodide](https://github.com/pyodide/pyodide). I've already attempted that for one of my [projects](https://stefnotch.github.io/quantum-sheet/) and it works reasonably well. 

Another option would be improving Sympy until it can reasonably be executed in the browser using Brython
https://github.com/sympy/sympy/issues/18889

**Comment by ivanistheone** [comment]
Yes, I've tested sympy in the [pyodide shell](https://pyodide.org/en/stable/usage/quickstart.html#alternative-example) and it works fine.

Typing `import sympy` leads to a O(10MB) async download of `mpmath` and `sympy`: 
<img width="1198" alt="loading async" src="https://user-images.githubusercontent.com/163966/125070639-3b4be080-e086-11eb-9513-e17c42a75cda.png">
After that every command I tested runs fine.

Overall this could be a way forward to remove or reduce the need for a backend for sympy live shell (and docs shell), for devices with good internet connection (where 10MB download is not an issue) and modern browsers (capable of running wasm).

**Comment by dynamicwebpaige** [comment]
Just tested SymPy with Pyodide, and experienced similar performance / latency to the SymPy Live implementation. It would be *awesome* to see Pyodide-enabled code samples in SymPy documentation.

**Comment by stefnotch** [comment]
I agree, this is definitely worth pursuing further! I suppose the absolutely simplest option would be serving a single, static `.html` file, [which loads Pyodide from the CDN](https://pyodide.org/en/stable/usage/index.html). However, that wouldn't be properly integrated into the rest of the project. Maybe someone who has more knowledge here could chip in on what the best way of going forward would be?

**Comment by asmeurer** [comment]
Moving SymPy Live and SymPy Gamma to Pyodide would be great if it works, and would save us a lot of money.

**Comment by ivanistheone** [comment]
I'm interested in contributing to and helping out any effort towards a pyodide-backed Gamma and Live.

I've been following this [thread](https://github.com/executablebooks/thebe/issues/465) and this [pull request](https://github.com/patrickmineault/thebe/commit/2f15a2a97607466b71c227ea1c6d501867205ccc) on the jupyter-book thiebe project, which I hope can be helpful as a "source of inspiration." JupyterLite is another code base worth watching [see [exectute](https://github.com/jupyterlite/jupyterlite/blob/52944e9cd21785291d0975a63f6723038efa9525/packages/pyolite-kernel/src/worker.ts#L155-L261) method].

**Comment by stefnotch** [comment]
So, taking another look at the code, if I understand it correctly
- [This code](https://github.com/sympy/sympy-live/tree/master/app) along with Sympy has to be executed in Pyodide on start
- [The "evaluate" call here](https://github.com/sympy/sympy-web-static/blob/a04cb7650583726f52a99013aab16bfcebfbd699/sympy-live/javascript/live-core.js#L899) has to be replaced with a call to Pyodide
- [The "autocomplete" call here](https://github.com/sympy/sympy-web-static/blob/a04cb7650583726f52a99013aab16bfcebfbd699/sympy-live/javascript/live-autocomplete.js#L144) should probably also be replaced with an appropriate call to Pyodide

Though, I suppose "replaced" might be the wrong word to use, since it might be worthwhile to keep the current App Engine backend around for now.

**Comment by jtpio** [comment]
Hi all,

Just landed here from https://github.com/executablebooks/thebe/issues/465.

`sympy` seems to be working fine with Pyodide. I haven't tried everything yet but so far it looks good.

Actually we have an example in the JupyterLite notebooks, if you want to try it out in the browser:

https://jupyterlite.readthedocs.io/en/latest/_static/retro/notebooks/?path=pyolite.ipynb

It also renders nicely just like with a regular Jupyter Notebook or with JupyterLab:

![image](https://user-images.githubusercontent.com/591645/137750797-ab6350f1-8a5f-465f-80a8-454f8e10cadf.png)

As well as in the code consoles: 

https://jupyterlite.readthedocs.io/en/latest/_static/retro/consoles/?path=/console-1-018f1666-a731-40f8-90dc-5196e5f6c23b

![image](https://user-images.githubusercontent.com/591645/137751353-91e00a1c-8d1a-441a-9369-8456dbf48e09.png)

**Comment by asmeurer** [comment]
I would expect if we switch to pyiodide the entire backend should be refactored. Right now it tries to persist variables to fake a session, since the App Engine doesn't persist a real session. But with pyiodide we can have a real persistent session on the client's computer. 

What I would love to see is a community supported Sphinx extension that lets you run examples in pyiodide, that is not tied to SymPy. That way we can have the same functionality we currently have in our docs, but without the burden of having to maintain it ourselves, which doesn't work so well given that most SymPy developers are not web developers. The same thing with the Live shell. We should look at converting SymPy Gamma to use pyiodide as well.

**Comment by eagleoflqj** [comment]
> I'm interested in contributing to and helping out any effort towards a pyodide-backed Gamma and Live.

Please take a look of the port of SymPy Gamma: https://github.com/eagleoflqj/sympy_beta
I plan to directly copy https://pyodide.org/en/stable/console.html to this project as a live shell

**Comment by ivanistheone** [comment]
@eagleoflqj This is very impressive work on [sympy_beta](https://github.com/eagleoflqj/sympy_beta). The Vue-fication makes the code very easy to read and understand. I suppose a similar approach for the live shell would be good, although we'd need a new "interactive shell" component like you said.

One thing to keep in mind is sympy-live shell is used both as a standalone webapp (https://live.sympy.org/) and also as a component in the docs, [example](https://docs.sympy.org/latest/tutorial/calculus.html) [include hook](https://github.com/sympy/sympy/blob/master/doc/ext/sympylive.py#L14-L28), and [src1](https://github.com/sympy/sympy-web-static/blob/a04cb7650583726f52a99013aab16bfcebfbd699/sympy-live/javascript/live-core.js) [src2](https://github.com/sympy/sympy-web-static/blob/a04cb7650583726f52a99013aab16bfcebfbd699/sympy-live/javascript/live-sphinx.js).

As you will see in [src1](https://github.com/sympy/sympy-web-static/blob/a04cb7650583726f52a99013aab16bfcebfbd699/sympy-live/javascript/live-core.js) [src2](https://github.com/sympy/sympy-web-static/blob/a04cb7650583726f52a99013aab16bfcebfbd699/sympy-live/javascript/live-sphinx.js) there is a lot of extra functionality bundled into the current live shell component, which would be really nice to keep if re-implementing in Vue.

There is also some logic in the Python code for evaluating expressions worth looking into
https://github.com/sympy/sympy-live/blob/master/app/handlers.py#L269-L275
https://github.com/sympy/sympy-live/blob/master/app/shell.py#L308-L321  although the pyodide shift will reduce the complexity A LOT since no more need for picklables and appengine session storage.

@eagleoflqj Could you please setup a barebones shell component (e.g. based on  https://pyodide.org/en/stable/console.html ) and share it so we can start experimenting with stuff?

First thing I will expriment is how to run python "middleware" on each exec/eval code, because these might be needed for post-processors like [PRINTERS](https://github.com/sympy/sympy-live/blob/master/app/constants.py#L60-L66) and [exception handlers](https://github.com/sympy/sympy-live/blob/master/app/handlers.py#L276-L299). Also probably for debugging too...

**Comment by ivanistheone** [comment]
@asmeurer said:
> What I would love to see is a community supported Sphinx extension that lets you run examples in pyiodide, that is not tied to SymPy. 

What you're describing is very close to this: https://github.com/executablebooks/sphinx-thebe if it had a pyodide backend.

example [book](https://minireference.github.io/noBSmath_lessons/tutorials/variables_and_expressions.html), use the Live Code option to enable
<img width="397" alt="Screen Shot 2021-12-31 at 1 18 42 PM" src="https://user-images.githubusercontent.com/163966/147835352-dd755694-61f6-4fb4-a3a7-ec0565a5c8c5.png">
thebe abstracts away the python execution "backend" and I'm sure the community will get a pyodide backend going soon enough (we just wait for it; [POC code](https://github.com/patrickmineault/thebe/commit/2f15a2a97607466b71c227ea1c6d501867205ccc) already available).

> That way we can have the same functionality we currently have in our docs, but without the burden of having to maintain it ourselves, which doesn't work so well given that most SymPy developers are not web developers.

Thebe backed runnable code in the docs would be a good option in terms of low maintenance, but break with the current UI:
- CURRENT of clicking on individual lines to execute them (which I don't think is too intuitive, but ppl might be used to it).
- NEW each code block is it's own "shell"

If we could have something like a thebe-shell component in additinon to the inline code blocks that would be the best, with sympy-live being powered by the same thebe-shell component.

**Comment by eagleoflqj** [comment]
> @eagleoflqj Could you please setup a barebones shell component (e.g. based on https://pyodide.org/en/stable/console.html ) and share it so we can start experimenting with stuff?

I just took the first step to integrate pyodide shell into sympy beta https://github.com/eagleoflqj/sympy_beta/commit/1ee37be6f2df33b4b87cc1ad66da4feea8977ba5
Currently when you switch between views the shell is reloaded, and if your input reaches the bottom it won't automatically scroll to a good view. I'll try to fix them, and maybe improve it by supporting multiple, minimize-able, restart-able terminal. But feel free to try this version.
As I directly used code from pyodide, this component has to be licensed no permissive than MPL-2.0 (and I choose to keep MPL-2.0). I don't know whether sympy live can benefit from it.

**Comment by jtpio** [comment]
FYI I opened https://github.com/sympy/sympy-live/pull/198 to switch to using JupyterLite for the REPL on https://live.sympy.org:

![image](https://user-images.githubusercontent.com/591645/152344748-f083ce00-f575-438d-9793-462058c1f68d.png)

JupyterLite comes with the Pyodide (currently at version 0.19.0) based Python kernel, and supports for rich output rendering and `IPython`.
